### ISSUE TITLE
fping responses misdelivered on FreeBSD

### ISSUE THREAD

**Issue by clarsen** [schweikert/fping#206]
On FreeBSD, the PIDs are as large as 99999, so if multiple fping processes are running there's a chance ident4 = PID & 0xffff will collide - which happens on monitoring hosts, like Nagios, that run fping often and simultaneously on a single IP address at a time.

If you have one IP up and one IP down and the two fping's PID ident4s collide, then the process pinging the down IP may see a reply from the up IP.

While the FreeBSD implementation of ping [calls connect()](https://github.com/freebsd/freebsd/blob/master/sbin/ping/ping.c#L644) which results in echo replies from undesired hosts being dropped before ever being delivered to the process, fping doesn't, so it has to rely on ICMP identifier to filter out the undesired responses.

It appears wait_for_reply() although it filters out ICMP identifier mismatch, it doesn't filter out on source IP address.

**Comment by clarsen** [comment]
bump on this issue - there's a pull request we've been running with for 18 months or so so far...

**Comment by eKrajnak** [comment]
I'm experiencing strange issue with fping 5.0 (bullseye repo) on Debian 11 (x64) and I think is colidate with this issue. Fping is used by Zabbix, we spawn 130 pinger processes simultaneously, usually busy for 40-50%, pinging 1k hosts every second (params -C10 -i0 -S [publicip]). Usually after VM reboot, random fping processes report random packet loss for random targets, even when I see icmp response in tcpdump. It's true that we ping a tens of permanently unavailable targets. This situation usually lasts few days, then disappear.

Is this bug affecting Linux distro also? Can we except fix for bullseye repos?
Thank you

**Comment by hmh** [comment]
Once it is fixed upstream, you *have* to file a Debian bug explaining the situation, pointing to this bug, and identifying the affected (buggy) versions of fping and the fixed version.  *Then*, it depends on whether the downstream Debian fping maintainer will prepare a debian stable update for fping (possibly backporting only this fix).

**Comment by gsnw** [comment]
@clarsen Do you happen to have an easy way to reproduce the error without using Nagios under FreeBSD?

**Comment by auerswal** [comment]
I would expect `fping` to print out the differing IP address, e.g.:

```
$ fping 192.0.2.47
 [<- 192.0.2.33]192.0.2.47 is alive
```

This is not actual `fping` output, but based on the output of pinging a multicast address:

```
$ fping 224.0.0.1
 [<- 172.20.10.1]224.0.0.1 is alive
```

The part in square brackets is sent to standard error, the rest to standard output. Is similar output observable in the Nagios and/or Zabbix use case?

This could also affect Linux, because PIDs on Linux can be larger than 65535.

I'd think that the `fping` default is useful behavior, i.e., I would not like to discard answers from a different IP address by default. I'd think that a new option that only accepts replies from the IP address the ICMP Echo was sent to could be useful for the monitoring use case. A test for this new option could ping a multicast address, e.g., `ff02::1` as used in `ci/test-01-basics.pl`.

**Comment by gsnw** [comment]
We can temporarily adopt clarsen's solution for FreeBSD only. It is included in the FreeBSD ports anyway.
At the moment I don't have any ideas for a nice solution. I'm still trying to understand the whole thing.

**Comment by auerswal** [comment]
I would not like to create this special case for FreeBSD, because I expect the problem to also affect GNU/Linux. As I have written above, I also dislike changing the default behavior of `fping` to ignore replies from source addresses that have not been added as targets *by default*.

I'd say that a new (long) option, e.g., `--strict-replies` (or `--check-source`), could be added to ignore reply packets with a source IP that differs from the IP address the request was sent to. This could look as follows:

```
$ fping 224.0.0.1 ; echo $?
 [<- 172.20.10.1]224.0.0.1 is alive
0
$ fping --strict-replies 224.0.0.1 ; echo $?
224.0.0.1 is unreachable
1
$ fping ff02::1 ; echo $?
 [<- fe80::abcd:…%enxabcd…]ff02::1 is alive
0
$ fping --strict-replies ff02::1 ; echo $?
ff02::1 is unreachable
1
```

The last example could be added as a CI test case, since `fping ff02::1` is already part of the tests. Using `fping 224.0.0.1` has not been reliable for me, i.e., in some networks it works as shown above, but in others it reports `224.0.0.1 is unreachable`, thus I have not attempted to add this to our CI.

I would expect that we might be able to use the existing code that verifies the IP address for implementing the new option, instead of adding a new verification loop, but I have not yet looked into the details. Otherwise, we could use the approach from @clarsen's patch extended with a new guard variable controlled by the new option.

We could also look into the identification issue a bit more. It might be possible to make use of the sequence number in addition to the identifier to avoid the reported issue. This could be seen as orthogonal to the `--strict-replies` option, i.e., adding the option would not inhibit improving reply identification. But I would expect this improvement to be more work than the new option.

I hope to find the time to look into adding a new option over the next weekend.

**Comment by auerswal** [comment]
@clarsen, does using `--check-source` avoid the problems you see with many `fping` instances running in parallel?

**Comment by clarsen** [comment]
sorry it took a while for me to re-reproduce the issue on my end to confirm the --check-source fix.  the tl;dr: this is my "repro" case and I confim the --check-source fix works.

1. Set ident4 in fping.c to `ident4 = ident6 = htons(0x8000 & 0xFFFF);` instead of `ident4 = ident6 = htons(getpid() & 0xFFFF);` to force the "colliding PID" scenario.

2. run a non-responsive ping in one process with a long enough timeout so that incoming responses from another source address will still be treated as valid.

`src/fping -t 5000 -c 5 192.168.100.100`

3. run a responsive ping in another process to produce responses

`src/fping -c 5 <any IP that responds>`

example of fail:
```
# src/fping -t 5000 -c 5 192.168.100.100
 [<- 131.243.5.1]192.168.100.100 : [0], 64 bytes, 940 ms (940 avg, 0% loss)
 [<- 131.243.5.1]192.168.100.100 : [1], 64 bytes, 948 ms (944 avg, 0% loss)
 [<- 131.243.5.1]192.168.100.100 : [2], 64 bytes, 948 ms (945 avg, 0% loss)
 [<- 131.243.5.1]192.168.100.100 : [3], 64 bytes, 944 ms (945 avg, 0% loss)
 [<- 131.243.5.1]192.168.100.100 : [4], 64 bytes, 999 ms (956 avg, 0% loss)

192.168.100.100 : xmt/rcv/%loss = 5/5/0%, min/avg/max = 940/956/999
```

example of using --check-source:
```
# src/fping -t 5000 -c 5 --check-source 192.168.100.100
192.168.100.100 : [0], timed out (NaN avg, 100% loss)
192.168.100.100 : [1], timed out (NaN avg, 100% loss)
192.168.100.100 : [2], timed out (NaN avg, 100% loss)
192.168.100.100 : [3], timed out (NaN avg, 100% loss)
192.168.100.100 : [4], timed out (NaN avg, 100% loss)

192.168.100.100 : xmt/rcv/%loss = 5/0/100%
```

**Comment by auerswal** [comment]
@clarsen: Thanks for testing and confirming that the new `--check-source` option works for you. :-)

**Comment by auerswal** [comment]
@eKrajnak: Your problem description seems to pertain to a different situation. You describe spurious *unreachable* results, but this issue pertains to spurious *reachable* results.

**Comment by K0nne** [comment]
Hello,

I can confirm this fix works to tackle phantom-reacheable results for offline devices.  
We run a icinga2 installation (2600 fping processes/min) on rhel8 and are affected by this issue.

**Comment by gsnw-sebast** [comment]
If the problem still exists, simply create a new ticket.
Judging by the comments, the problem has been solved.
Thanks for the feedback
